network-core/src/gossip.rs

```diff
@@ -1,25 +1,33 @@
 use chain_core::property::{Deserialize, Serialize};
 
 #[derive(Clone, Debug, PartialEq, PartialOrd)]
-pub struct NodeId([u8; 16]);
+pub struct Id(pub [u8; 16]);
 
-pub enum NodeIdError {
+pub enum IdError {
     InvalidSize(usize),
 }
 
-impl NodeId {
-    pub fn from_slice(slice: &[u8]) -> Result<Self, NodeIdError> {
+impl Id {
+    pub fn from_slice(slice: &[u8]) -> Result<Self, IdError> {
         if slice.len() != 16 {
-            return Err(NodeIdError::InvalidSize(slice.len()));
+            return Err(IdError::InvalidSize(slice.len()));
         }
         let mut buf = [0; 16];
         buf[0..16].clone_from_slice(slice);
-        Ok(NodeId(buf))
+        Ok(Id(buf))
     }
 
     pub fn to_bytes(&self) -> Vec<u8> {
         self.0[..].into()
     }
 }
 
-pub trait Gossip: Serialize + Deserialize {}
+pub trait Gossip: Serialize + Deserialize {
+    type Id: Sized;
+    type Node;
+
+    fn to_node_id(id: &Self::Id) -> Id;
+    fn from_node_id(id: &Id) -> Self::Id;
+
+    fn from_nodes(iter: Box<Iterator<Item = Self::Node>>) -> Self;
```

Comment:
This could be generic rather than forcing an allocation:

```rust
    fn from_nodes<I>(iter: I) -> Self
    where I: IntoIterator<Item = Self::Node>
    {
    }
```